rbd: fix bug with missing supported_features

pkalever opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

* continue running rbd driver when /sys/bus/rbd/supported_features file is
missing, do not bail out.
* handle when krbdFeatures is zero


## Related issues ##

Fixes: #2678 

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

Commit:
rbd: fix bug handling GetKrbdSupportedFeatures()

continue running rbd driver when /sys/bus/rbd/supported_features file is
missing, do not bailout.

Fixes: #2678
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -265,11 +265,16 @@ func HexStringToInteger(hexString string) (uint, error) {
 // isKrbdFeatureSupported checks if a given Image Feature is supported by krbd
 // driver or not.
 func isKrbdFeatureSupported(ctx context.Context, imageFeatures string) bool {
+	supported := true
+	// return true when /sys/bus/rbd/supported_features is absent
+	// ex: kernel < v4.11.0
+	if _, err := os.Stat(krbdSupportedFeaturesFile); errors.Is(err, os.ErrNotExist) {
+		return supported
+	}
```
if `krbdFeatures` is 0 means we are not able to get the `supported_featurs`  we don't need to do stat again and again?

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -264,7 +264,12 @@ func HexStringToInteger(hexString string) (uint, error) {
+func isKrbdFeatureSupported(ctx context.Context, imageFeatures string) (bool, error) {
+	// return false when /sys/bus/rbd/supported_features is absent
+	// ex: kernel < v4.11.0
+	if krbdFeatures == 0 {
```
if krbdFreatues is 0 just return true? as we don't need to support `tryOtherMounters` if we are not able to fetch the rbd features from the supported_features file.

Madhu-1 replied:
@pkalever looks like this is not addressed. can you please check it again?

pkalever replied:
@Madhu-1 
Sorry! After all, I still strongly feel `isKrbdFeatureSupported()` should return an error when the file is not found instead of faking true/false. Can you just think over it once again, please?

nixpanic commented on internal/rbd/rbd_util_test.go:
```diff
@@ -346,8 +347,10 @@ func TestIsKrbdFeatureSupported(t *testing.T) {
-			supported := isKrbdFeatureSupported(ctx, tc.featureName)
-			if supported != tc.isSupported {
+			supported, err := isKrbdFeatureSupported(ctx, tc.featureName)
+			if err != nil && !errors.Is(err, os.ErrNotExist) {
```
would be good to have a comment why `os.ErrNotExist` is ignored

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -241,8 +241,12 @@ func populateRbdVol(
+	isFeatureExist, err := isKrbdFeatureSupported(ctx, strings.Join(rv.ImageFeatureSet.Names(), ","))
+	if errors.Is(err, os.ErrNotExist) {
+		// do not fallback if /sys/bus/rbd/supported_features is absent
+		isFeatureExist = true
```
so, if the file is missing, assume the feature is present? Is that not problematic with new features on old RBD drivers?

pkalever replied:
We do not have a way to find what features are present when `/sys/bus/rbd/supported_features` is absent, indeed that is the reason why `/sys/bus/rbd/supported_features` file was introduced.

Are you suggesting that we should fallback to NBD if the `/sys/bus/rbd/supported_features` is absent?

nixpanic replied:
Just wondering what the most appropriate default would be. If `/sys/bus/rbd/supported_features` is missing, the driver is old, so fewer features would be supported. Consider having a list of supported features of the last driver version that does not provide this API.

pkalever replied:
@Madhu-1 what is your take on this? Instead should we fallback to NBD incase if the krbd driver is old then?

Madhu-1 replied:
but it's not expected behavior for existing clusters.

pkalever replied:
Then I suppose this will be a breaking change and should be introduced in 3.7.0 only. Thoughts?

Madhu-1 replied:
IMO this will be good only if nbd is stable and can be used in production for 3.7.0.

nixpanic replied:
do you know what features are supported by the most recent driver that does not have the `/sys/bus/rbd/supported_features` file? In case the file is missing, you could check if it is a feature that is supported long (enough).

pkalever replied:
Upstream /sys/bus/rbd/supported_features is part of Linux kernel v4.11.0
RBD_FEATURE_LAYERING is added around Linux v3.8,
RBD_FEATURE_STRIPINGV2 is added around v3.10,
RBD_FEATURE_EXCLUSIVE_LOCK is added around v4.9, and
RBD_FEATURE_DATA_POOL around v4.11

nixpanic replied:
The reporter of the issue used Linux 4.4.0. I guess some Debian version might still be on such an old kernel?

As you have the features and their kernel versions, you can use that (and ignore backports). In case `/sys/bus/rbd/supported_features` exists, use that as authoritative source.

pkalever replied:
@Madhu-1 Are you okay with ignoring backports too and basically with Niels comment above?

Madhu-1 replied:
Yes, should be fine :+1:

nixpanic replied:
Waiting on an update for this, marking as required change.

pkalever replied:
Done now, PTAL.

Madhu-1 commented:
@mergifyio rebase

pkalever commented:
@Mergifyio rebase

Madhu-1 commented:
/retest ci/centos/upgrade-tests-rbd

Commit:
rbd: prepare krbd feature attrs if supported_features file is absent

Upstream /sys/bus/rbd/supported_features is part of Linux kernel v4.11.0
Prepare the attributes and use them in case if
/sys/bus/rbd/supported_features is missing.

Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

pkalever commented on internal/rbd/rbd_util.go:
```diff
@@ -193,28 +203,94 @@ type migrationVolID struct {
+		return strconv.FormatInt(rbdStripingV2SupportAttr, 16), nil
+	} else if util.CheckKernelSupport(release, krbdLayeringSupport) {
+		return strconv.FormatInt(rbdLayeringSupportAttr, 16), nil
+	}
```
@Madhu-1 I don't see how to convert this if-Else to Switch as suggested by linter. Can I just add a nolint here? thoughts?

Madhu-1 replied:
Yes should be fine :+1:  Please add a reason for adding nolint.

pkalever replied:
Sure. Thanks!

pkalever commented:
@Madhu-1 @nixpanic PTAL.

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

pkalever commented:
/retest ci/centos/upgrade-tests-cephfs

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

yati1998 reviewed (dismissed):
LGTM

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -193,28 +203,95 @@ type migrationVolID struct {
+)
+
+// prepareKrbdFeatureAttrs prepare krbd fearure set based on kernel version.
+// Minimum kernel version should be 3.6, else it will return error.
```
the minimal kernel is 3.6 or 3.8?

pkalever replied:
Typo, yes 3.8

Madhu-1 reviewed (commented):
I will let @nixpanic do the final review.

pkalever commented:
@Mergifyio rebase

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
Diff b/w previous version and this version.
V2: Updated this PR resolving a merge conflict in noderserver.go because of https://github.com/ceph/ceph-csi/pull/3080

pkalever commented:
/retest ci/centos/k8s-e2e-external-storage/1.22

pkalever commented:
/retest ci/centos/mini-e2e/k8s-1.23

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -79,6 +79,16 @@ const (
 	// krbd attribute file to check supported features.
 	krbdSupportedFeaturesFile = "/sys/bus/rbd/supported_features"
+
+	rbdFeatureLayering      = (1 << 0) // RBD_FEATURE_LAYERING
```
these constants/#defines are already available in go-ceph:

- https://github.com/ceph/go-ceph/blob/master/rbd/features.go
- https://github.com/ceph/go-ceph/blob/master/rbd/features_nautilus.go

Better use those.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -193,28 +203,95 @@ type migrationVolID struct {
+			SubLevel:     0,
+			ExtraVersion: 0,
+			Distribution: "",
+			Backport:     false,
```
you can leave the `ExtraVersion`, `Distribution` and `Backport` attributes out, those are default values anyway and will make the code little shorter.

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -193,28 +203,95 @@ type migrationVolID struct {
+
+// prepareKrbdFeatureAttrs prepare krbd fearure set based on kernel version.
+// Minimum kernel version should be 3.8, else it will return error.
+// nolint: gocritic // cannot convert if-else to switch.
```
please do that, it is a minimal change and prevents needed cleanup later

pkalever replied:
@nixpanic 
I thought we cannot convert this if-else to switch, at least it's not straightforward to me. Can you recheck once?

nixpanic replied:
I guess, something like this:

```go
switch {
    case util.CheckKernelSupport(release, krbdDataPoolSupport):
        return librbd.FeatureDataPool, nil
    case ...:
        return ...
}
```

pkalever replied:
Oh, a switch without a value is possible!! I totally forget that :) Thanks!

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -193,28 +203,95 @@ type migrationVolID struct {
+// prepareKrbdFeatureAttrs prepare krbd fearure set based on kernel version.
+// Minimum kernel version should be 3.8, else it will return error.
+// nolint: gocritic // cannot convert if-else to switch.
+func prepareKrbdFeatureAttrs() (string, error) {
```
why return a string, and not an int?

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -79,6 +79,16 @@ const (
+	rbdFeatureExclusiveLock = (1 << 2) // RBD_FEATURE_EXCLUSIVE_LOCK
+	rbdFeatureDataPool      = (1 << 7) // RBD_FEATURE_DATA_POOL
+
+	rbdLayeringSupportAttr      = (rbdFeatureLayering)
```
you could use go-ceph's `FeatureSet` for this

nixpanic commented on internal/rbd/rbd_util.go:
```diff
@@ -239,9 +316,19 @@ func GetKrbdSupportedFeatures() (string, error) {
+			return "", err
+		}
+
+		return attrs, nil
```
it would be cleaner to do the `strconv.FormatInt()` here, less repeated code in `prepareKrbdFeatureAttrs()`

nixpanic commented on internal/rbd/nodeserver.go:
```diff
@@ -241,8 +241,13 @@ func populateRbdVol(
+	if errors.Is(err, os.ErrNotExist) {
+		// do not fallback if /sys/bus/rbd/supported_features is absent and we are
+		// not in a position to prepare krbd feature attributes
+		isFeatureExist = true
```
should this default to `true`? I would expect that if the kernel is too old, and the file does not exist, it is assumed the feature is not available either.

pkalever replied:
@nixpanic are you saying, If the kernel is too old and the file doesn't exist and hence we set krbdFeatures to 0, so fall back to rbd-nbd?

nixpanic replied:
not sure how you mean that. But I read the code as "if no features can be detected, assume the feature is supported". And I think that is wrong, and would expect the kernel to not support the feature.

nixpanic replied:
Still waiting for a reply on this, thanks!

pkalever replied:
@nixpanic 
Can you check now! Thanks.

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22
```
May 11 14:03:25.573: FAIL: backend images not matching kubernetes resource count,image count 23 kubernetes resource count 21

```

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Commit:
rbd: handle when krbdFeatures is zero

krbdFeatures is set to zero when kernel version < 3.8, i.e. in  case where
/sys/bus/rbd/supported_features is absent and we are unable to prepare
the krbd attributes based on kernel version.

When krbdFeatures is set to zero fallback to NBD only when autofallback
is turned ON.

Fixes: #2678
Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

humblec commented:
@pkalever linter is faililng..

humblec commented on internal/rbd/nodeserver.go:
```diff
@@ -241,8 +241,14 @@ func populateRbdVol(
-		!isKrbdFeatureSupported(ctx, strings.Join(rv.ImageFeatureSet.Names(), ",")) {
+	isFeatureExist, err := isKrbdFeatureSupported(ctx, strings.Join(rv.ImageFeatureSet.Names(), ","))
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		log.ErrorLog(ctx, "failed checking krbd features: %v", err)
```
do we need to listout the feature as well?

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -192,28 +192,83 @@ type migrationVolID struct {
-	librbd.FeatureNameDeepFlatten: {
-		needRbdNbd: false,
-	},
+var (
```
@pkalever can we also document the justifying link as source code comment for below version mapping as referecnce ?

pkalever replied:
There is nothing as such, I had to go to linux tree and find it myself :-(

humblec commented on internal/rbd/rbd_util.go:
```diff
@@ -238,9 +293,19 @@ func GetKrbdSupportedFeatures() (string, error) {
+		}
+		attrs, err := prepareKrbdFeatureAttrs()
+		if err != nil {
+			log.ErrorLogMsg("preparing krd feature attributes failed, %v", err)
```
krd/krbd

pkalever commented:
/restest ci/centos/mini-e2e-helm/k8s-1.21
/restest ci/centos/mini-e2e-helm/k8s-1.22
/restest ci/centos/mini-e2e-helm/k8s-1.23
```
May 13 10:31:50.222: FAIL: backend images not matching kubernetes resource count,image count 11 kubernetes resource count 10
```

pkalever commented:
/restest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/restest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
/restest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

pkalever commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.23

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21
